### ISSUE TITLE
COF and SOD huge bugs?

### ISSUE THREAD

**Issue by dcabanas** [yzhao062/pyod#241]
Hi all. Something is happening that I cannot understand, I'm doing hyperparameter tuning to all proximity based algorithms with RandomSearchCV form sklearn and only COF and SOD have problems. What is happening is the following:  the constructor parameters are all None even though I'm passing values to it. I tried to change the code of SOD and hardcoded the values of the constructor and I'm still getting NoneType. Please help.

**Comment by John-Almardeny** [comment]
Can you please provide a reproducible example?

**Comment by dcabanas** [comment]
I have this code in a file called hypertuning.py in the "notebooks" folder, you can put it there and runs smoothly. I will explain the behaviour and it's easier if you run it: basically for all the algorithms I'm trying to optimize the parameters and the "print(clf)" line tells me, after fitting, the parameters that RandomSearch is using for the current model. For COF and SOD the parameters are None no matter what I do. This code saves the best parameters in best_params.txt file in the current folder, "notebooks".

`import warnings
warnings.filterwarnings("ignore")

from sklearn.model_selection import RandomizedSearchCV

import numpy as np
import pandas as pd
from pandas import read_excel

from pyod.models.lof import LOF
from pyod.models.cof import COF
from pyod.models.cblof import CBLOF
from pyod.models.hbos import HBOS
from pyod.models.knn import KNN

from pyod.models.sod import SOD


BASE_DIR = '../../'

datasets = [
    'MED-WTG1-DATA.xlsx',
    'MED-WTG2-DATA.xlsx',
    'MED-WTG3-DATA.xlsx',
    'MED-WTG4-DATA.xlsx',
    'MED-WTG5-DATA.xlsx',
    'MED-WTG6-DATA.xlsx',
    'MED-WTG7-DATA.xlsx',
    'MED-WTG8-DATA.xlsx',
]

lof_random = {
    'n_neighbors': np.arange(1, 101),
    'algorithm': ['ball_tree', 'kd_tree', 'auto'],
    'leaf_size': np.arange(20, 41),
    'metric': ['minkowski', 'euclidean']
}

cof_random = {
    'n_neighbors': np.arange(1, 101)
}

cblof_random = {
    'n_clusters': np.arange(2, 101),
    'alpha': np.arange(0.5, 1, 0.05)
}

hbos_random = {
    'n_bins': np.arange(1, 101)
}

knn_random = {
    'n_neighbors': np.arange(1, 101),
    'algorithm': ['ball_tree', 'kd_tree', 'auto'],
    'leaf_size': np.arange(20, 41),
    'metric': ['minkowski', 'euclidean'],
}

sod_random = {
    'alpha': np.arange(0.5, 1, 0.05)
}

classifiers_params = {
    'Local Outlier Factor (LOF)': [LOF(), lof_random],
    'Connectivity-Based Outlier Factor (COF)': [COF(), cof_random],
    'Clustering-Based Local Outlier Factor (CBLOF)': [CBLOF(), cblof_random],
    'Histogram-based Outlier Score (HBOS)': [HBOS(), hbos_random],
    'K Nearest Neighbors (KNN)': [KNN(), knn_random],
    'Average K Nearest Neighbors (AvgKNN)': [KNN(method='mean'), knn_random],
    'Median K Nearest Neighbors (MedKNN)': [KNN(method='median'), knn_random],
    'Subspace Outlier Detection (SOD)': [SOD(), sod_random] 
}

with open('best_params.txt', 'w', encoding='utf8') as pepi:
    for j in range(len(datasets)):

        data_file = BASE_DIR + datasets[j]
        data_name = data_file[len(BASE_DIR):-5]
        data = read_excel(data_file).to_numpy()[:,[2,3,5,7,10,11]]
        data = np.array(data, dtype=np.float32)

        print("\n... Processing", data_name, '...')
        pepi.write(data_name + '\n')

        X = data[:,:-1]
        y = data[:,-1]

        for clf_name, clf in classifiers_params.items():
            print("... Hyper tuning", clf_name, "...")
            print(clf)

            cv = [(slice(None), slice(None))]
            rf_random = RandomizedSearchCV(estimator=clf[0], param_distributions=clf[1],
                                            n_iter=10, cv=cv, n_jobs=-1, scoring='roc_auc', verbose=2)
            rf_random.fit(X, y)
            best_params_dict = rf_random.best_params_
            best_score = round(rf_random.best_score_, 2)

            print('\n Best params for ' + clf_name)
            print(best_params_dict)
            print()
            pepi.write('Best params for ' + clf_name + ' with best AUC ' + str(best_score) + '\n')
            for k,v in best_params_dict.items():
                pepi.write(k + ': ' + str(v) + '\n')
            pepi.write('\n')

`

**Comment by John-Almardeny** [comment]
Hi there,
I just fixed the bug. It is just a naming inconvenience with `Scikit-learn` conversion.
It is all sorted out now, I tried your code and it's working.
I opened a [Pull Request](https://github.com/yzhao062/pyod/pull/242) and waiting for Yue @yzhao062 to merge it.

Yahya.

**Comment by dcabanas** [comment]
I just saw the changes, thank you so much. Shall I close this issue or one of the contributors?

**Comment by dcabanas** [comment]
I just saw the changes, thank you so much. Shall I close this issue or one of the contributors closes it?

**Comment by John-Almardeny** [comment]
No problem. You're welcome. Yue shall close it once PR is merged.

Yahya.

**Comment by yzhao062** [comment]
Hey guys. I have another thought here...
The reason it is called n_neighbors_ is indeed following sklearn convention given the training samples are not enough...see an example of LOF here: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.html

It is possible to add some additional lines to have:
self.n_neighbors = n_neighbors, so your printing will be running ok. The current version is actually running correctly, while the true parameter value is stored in n_neighbors_ instead.

Thoughts...?

**Comment by dcabanas** [comment]
Hi,
I don't have the knowledge nor the sensibility in these matters like you guys but I can give an honest oppinion. The idea of following sklearn convention sounds good and I would rather prefer that you add those lines since the variable I'm using is indeed called n_neighors. Another thing that I forgot to mention that is also weird for me is the fact: I also ran the COF and SOD examples in the "examples" folder, and I printed the model after the "fit" call and n_neighbors are indeed None again but the example runs smootlhy. Right now, as a temporary solution maybe, I will use the changes that @John-Almardeny did since I have work to do. 

Thanks so far for all the help guys.

**Comment by dcabanas** [comment]
Sorry to bother guys but I already changed the files with the respective changes that @John-Almardeny did, on my local repo, tried to run without changing the code I posted here and still get the same error. I really can't figure out what is going on.

![Capture](https://user-images.githubusercontent.com/27977389/97105934-650ad600-16b6-11eb-9460-c2865e1b9ef5.PNG)

**Comment by John-Almardeny** [comment]
Change `n_jobs` to 1.

**Comment by dcabanas** [comment]
Just did and tried with other values for n_jobs for my RandomizedSearch and still the same error. Did I forget to change anything? I'm pretty I copied everything you changed.

**Comment by dcabanas** [comment]
Ok guys I found the problem! Since I'm running my project on a conda environment, even after I did your changes I need to run again "pip install ." to update my conda environment. Everything is working now.
I still think the changes you did are necessary. Sorry for the inconvenience.
